### ISSUE TITLE
[analyzer] Flaky property?

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#46206]
This issue reports on a phenomenon that I've known about for quite a while, but I haven't seen this phenomenon in small examples even though I tried several times to cut out a tiny core program to get a tight reproduction.

So now I'll report the issue anyway, based on the situation where the phenomenon occurs all the time; it involves code generation with the package reflectable. There is a small script below which will take the necessary steps. I hope this can serve as a starting point for detecting why the analyzer can report certain properties of `Elements` in a flaky manner.

The issue is that the getter `redirectedConstructor` of `ExecutableElement` returns different results at different times when executed on the exact same input, as if it isn't (always) initialized, even though there are no other reasons to believe that something hasn't been initialized.

Concretely, this means that we can generate code for the reflectable tests and run a specific test, and it succeeds; then we can generate code once more (nothing else changes) and run that test again, and now it fails. The failure occurs because a specific constructor no longer returns the correct value from `isRedirectingConstructor` of `MethodMirror` (from `package:reflectable/mirrors.dart`), which is implemented using the above mentioned getter.

A similar phenomenon can occur with the property `VariableMirror.isFinal` which is implemented using `VariableElement.isFinal`.

The issues seem to be connected to the checkout (two different copies on disk on my desktop with byte-equal test files behave differently), so maybe the ordering of code generation steps plays a role, which may come up because two directories can hold the same files in a different order (and perhaps the analyzer keeps some state from code generation step to code generation step).

```
> ... # Generate code
> for n in `seq 5`; do dart test; done
00:07 +496: All tests passed!                                                                                                                                                                                                                                                                                                                                             
00:07 +496: All tests passed!                                                                                                                                                                                                                                                                                                                                             
00:08 +496: All tests passed!                                                                                                                                                                                                                                                                                                                                             
00:08 +496: All tests passed!                                                                                                                                                                                                                                                                                                                                             
00:08 +496: All tests passed!                                                                                                                                                                                                                                                                                                                                             
> ... # Generate code again
> for n in `seq 2`; do dart test; done
00:07 +417 -1: test/declarations_test.dart: MethodMirror properties [E]                                                                                                                                                                                                                                                                                                   
  Expected: true
    Actual: <false>
  
  package:test_api                   expect
  test/declarations_test.dart 148:5  main.<fn>
  
00:07 +417 -1: loading test/declarations_test.dart                                                                                                                                                                                                                                                                                                                        
Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'pub run test --chain-stack-traces'.
00:07 +495 -1: Some tests failed.                                                                                                                                                                                                                                                                                                                                         
00:07 +417 -1: test/declarations_test.dart: MethodMirror properties [E]                                                                                                                                                                                                                                                                                                   
  Expected: true
    Actual: <false>
  
  package:test_api                   expect
  test/declarations_test.dart 148:5  main.<fn>
  
00:07 +417 -1: loading test/declarations_test.dart                                                                                                                                                                                                                                                                                                                        
Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'pub run test --chain-stack-traces'.
00:08 +495 -1: Some tests failed.                                                                                                                                                                                                                                                                                                                                         
```

The success/failure is determined by generated code, so it will succeed every time after a "good" code generation step, and it will fail every time after a "bad" step (and the latter is a code generation step where `ExecutableElement.redirectedConstructor` returned null even though the given constructor _is_ redirecting).

Here is a script that will give rise to a successful run of all tests at times, and fail in one or two tests at other times, on Linux:

```bash
#!/bin/bash --norc

if [ -e reflectable.dart ]; then
  echo 'Please run this script in a location where a'
  echo 'directory named `reflectable.dart` can be created.'
  exit -1
fi

git clone https://github.com/google/reflectable.dart.git

cd reflectable.dart/test_reflectable/tool
make clean get
cd ..
while true; do
  echo "-------------------------------------------------- Build"
  dart run build_runner clean
  dart run build_runner build --delete-conflicting-outputs
  echo "-------------------------------------------------- Run"
  dart test
done
```

**Comment by scheglov** [comment]
Could you please guide me some more?
I was able to run your script, and I see that sometimes there are 1, sometimes 2, and sometimes 3 failing tests.
But I don't really know what I am looking at.
Is there some code that generated differently depending on behavior of the analyzer?
Where is this generated code?
Where is the generator code that does it? Even better, where is the pivotal branch in this generator that causes one or another code to be generated?

With this, I would be much easier for me too know which properties in the analyzer to debug-output and track.

**Comment by scheglov** [comment]
OK, you might have already provided enough hints, I see that when I log in `ConstructorElementImpl.redirectedConstructor`, in rare cases (1 out of 7, in the same run) the result is `null` for `A.redirecting()`. I will look further, but wanted to let you know that you don't need to spent time on this.

**Comment by scheglov** [comment]
I think this fixes the issue.
https://dart-review.googlesource.com/c/sdk/+/209200

Basically we were recording `redirectedConstructor` during resolution for non-const constructors, because the idea was that you don't need it, if you don't evaluate constants. But probably if it is something exposed through the element model, and you can ask just the element model, it is not a good idea to mutate it.

**Comment by eernstg** [comment]
Actually, the behavior still exists when I use the bleeding edge version of the analyzer. So I may still need to try to extract a minimal example (even though I suspect that it might only happen in a not-so-minimal program...).

**Comment by scheglov** [comment]
Strange.
OK, let me know if there are any more details.

**Comment by eernstg** [comment]
I'd love to provide more details, but perhaps we should arrange for a video meeting where I can pick your brain about how to do it? ;-) Perhaps Tuesday?

**Comment by scheglov** [comment]
Sure, Tuesday will work. Schedule something. I'm on the east coast currently.

**Comment by eernstg** [comment]
It might be working after all!

I ran a test like the one in the script [here](https://github.com/dart-lang/sdk/issues/46206#issue-908294239) on Aug 6th, 5:45pm CEST, using the analyzer from SDK commit 003e79752c9cd4e57de92018c208d48d3f69aa11 (several commits after https://github.com/dart-lang/sdk/commit/39178e6b130ecc764b903c71865ecc22bc5adede, so it _should_ include your fix), and noticed several failures in `test/declarations_test.dart` at line 148 and/or in `test/legacy/declarations_test.dart` at line 149 (along with a lot of successful runs).

However, today I did `dart pub upgrade`, now using the analyzer from SDK 020f21261ab531f56f6fddb5d2fe42a0f6928bbb (yesterday afternoon, CEST), and there have not been any failures at all.

So I'll close this again. Thanks!
